"""
In progress. disregard.
"""
from Display import *
from Maths import *
from Microarray_handler import *
from Preprocessing import merge
from SNR_tools import signal_noise_pangenomic
from Utils import *

"""
df = pd.read_csv("ressources/MMACHC1.csv", sep=";", dtype={"Chromosome": str})

df_merged = merge(df, df2)
"""
df2 = pd.read_csv("ressources/Marker1.csv", sep=";")

df3 = pd.read_csv("ressources/EWAS_regression_aging_450K.csv", sep=";")
df3 = merge(df3, df2)
df3 = df3.dropna(subset=['Position', 'Chromosome'])

df3['Chromosome'] = df3['Chromosome'].astype(str).str.replace('.0', '', regex=False).str.strip()

savgol_smoothing(df3,"log10",17,2)

snr_df = signal_noise_pangenomic(df3, "smooth_result", window_size=5)
snr_filtered = snr_df[snr_df["smooth_result"] > 18]

genes = snr_filtered["Gene Name"].dropna().unique()

genes_sorted = sorted(genes)
print("Liste des gÃ¨nes Ã  haut SNR (smooth_result > 11):")
for gene in genes_sorted:
    print(gene)

plot_manhattan(df3, "smooth_result")


"""
df = read_sdrf("ressources/E-GEOD-30870.sdrf.txt")

newborns_df, nona_df = build_methylome_dataframes(
        "ressources/E-GEOD-30870.sdrf.txt",
        "ages",
    )

stats = run_methylation_ttest(
    newborns_df, nona_df,
    name1="Newborns",
    name2="Nonagenarians")

stats.to_csv("ressources/methylation_ttest_results.csv")

print(stats.head())

plot_volcano(stats, title="EWAS Newborns vs Nonagenarians",
             p_thresh=0.05, effect_thresh=0.10)


"""

#plot_manhattan2(df3,snr_filtered)


#mean_smoothing(df3,"log10",5)
#plot_manhattan(df3, "smooth_result")

#x = bootstrap_mean_with_CI(df_merged, 'MMACHC','T_log10_P')
#plot_bootstrap_mean_with_CI(x)

#d = bootstrap_distribution(df_merged)
#plot_bootstrap_violon(d)


#s = mean_smoothing(df_merged, 5)
#snr_df = signal_noise_pangenomic(s)
#snr_filtered = snr_df[snr_df["Signal_Noise_Ratio"] > 6]
#plot_manhattan2(s,snr_filtered,3)
